flypipe/node.py

```diff
@@ -221,51 +222,66 @@ def __call__(self, *args):
         return self.function(*args)
 
     def run(  # pylint: disable=too-many-arguments
-        self,
-        spark=None,
-        parallel=None,
-        inputs=None,
-        pandas_on_spark_use_pandas=False,
-        parameters=None,
+            self,
+            spark: SparkSession = None,
+            parallel: bool = None,
+            inputs: dict = None,
+            pandas_on_spark_use_pandas: bool = False,
+            parameters: dict = None,
+            cache: dict = None,
     ):
+        if spark:
+            if not isinstance(spark, SparkSession):
+                raise TypeError(f"spark is not of type pyspark.sql.SparkSession")
+
         if not inputs:
             inputs = {}
 
-        provided_inputs = {node.key: df for node, df in inputs.items()}
-        self._create_graph(
-            list(provided_inputs.keys()), pandas_on_spark_use_pandas, parameters
-        )
+        run_context = RunContext(spark=spark,
+                                 parallel=parallel,
+                                 provided_inputs=inputs,
+                                 pandas_on_spark_use_pandas=pandas_on_spark_use_pandas,
+                                 parameters=parameters,
+                                 cache_modes=cache)
+
+        self._create_graph(run_context)
 
-        if provided_inputs is None:
-            provided_inputs = {}
+        # re-create graph again with nodes caches

```

Comment:
I initially had implemented what you coded in `process_transformation_with_cache`. However I realised it was not the complete solution.

I forgot to change the comment `# re-create graph again with nodes caches`. In fact it creates the graph only once and then only load the caches for nodes with RunStatus.CACHED

**Cause #1 Writing cache within process_transformation**

It will write the cache with a dataframe as it came from the `self.process_transformation(spark, requested_columns, run_context, **inputs)` without removing columns defined in output schema

Example:

Create a node implements an output schema that outputs column 1 and 2. However the output of this node outputs a dataframe with column 1,  2 and 3. 
when writing the cache (i.e. a spark table) it will save with col 1, 2 and 3 when in fact it should save with 1 and 2 as defined in the output schema.

For this reason it is writing the dataframe after added to NodeResult as it will select appropriate columns ([L369](https://github.com/flypipe/flypipe/blob/436ec556add87e97c3f0e58496ce547490aa52f4/flypipe/node.py#L369) and [L294](https://github.com/flypipe/flypipe/blob/436ec556add87e97c3f0e58496ce547490aa52f4/flypipe/node.py#L294)




**Cause #2 Reading cache within process_transformation**

The decision of pre-loading all caches in `run_context.update_provided_inputs(self.node_graph.load_caches())` [line 250](https://github.com/flypipe/flypipe/blob/436ec556add87e97c3f0e58496ce547490aa52f4/flypipe/node.py#L250). 
Is that nodes with existent cache have to have RunStatus == SKIP so the graph html will show that these nodes have been skipped. However if they are skipped, this node will not be selected to run when [getting execution graph](https://github.com/flypipe/flypipe/blob/436ec556add87e97c3f0e58496ce547490aa52f4/flypipe/node.py#L256) and it will not run the method `process_transformation_with_cache`.

the alternative would to leave these nodes with `RunStatus.CACHED`, however the downstream impacts in the could would be greater as we would have to change ` execution_graph.get_runnable_transformations()`, `self.node_graph.get_execution_graph(run_context)` and the frontend html to deal with Cached status as skipped as well.

the already implemented logic `provided_inputs` deals with all of it. so I decided to pre load the caches only once (clearer where we are loading the caches ([run_context.provided_inputs = self.node_graph.load_caches()](https://github.com/flypipe/flypipe/blob/436ec556add87e97c3f0e58496ce547490aa52f4/flypipe/node.py#LL250C9-L250C68)), we keep all nodes that will not run with run status skip.

In this way we achieve the same behaviour, without having to deal with 2 inactive status SKIP and CACHED in multiple places.


I think the question we want to answer is:

makes sense to have read cache within `process_transformation_with_cache` considering that we have to write the cache outpuside using node result and also dealing with RunStatus.CACHED in other piece of logic (including frontend)?

Reply:
@ChrisMarwick What about if we keep `process_transformation_with_cache` but include NodeResult inside it to write the cache?

Also what do you think if we get rid of outputs from `def _run_sequential(self, run_context: RunContext, execution_graph, outputs)`, and let `process_transformation_with_cache` to load from node_run_context.provided_input or cache?

If we decided to go this way we have to create a new RunStatus.PROVIDED_INPUT and consider as runnable nodes with status `PROVIDED_INPUT`, `ACTIVE` and `CACHED`.

Reply:
Yep, this makes sense and I fully agree with the second comment. I think if you move around the function calls a bit you could support what you're trying to do with running df filtering inside dataframe wrapper before caching the raw dataframe, but have it happen within process_transformation_with_cache. 

Getting rid of outputs in favor of the new runtime context would be great- now that we have a context object with all the runtime stuff we should be able to box everything we need for runtime in there including outputs. It seems like a small change but I think it will help a lot with reading the code, for everything to do runtime to be boxed in one place. 

I think adding a new RunStatus should not be too bad- searching through the code I can only see a few uses of it in graph_html, a few tests and node_graph itself.